contracts/Controller.sol

```diff
@@ -124,38 +128,88 @@ contract Controller is Ownable {
      * @param _otoken The address of the relevant oToken.
      * @return A boolean which is true if and only if the price is finalized.
      */
-    // function isPriceFinalized(address _otoken) external view returns (bool) {
-    // }
+    function isPriceFinalized(address _otoken) external view returns (bool) {
+        address oracleModule = AddressBookInterface(addressBook).getOracle();
+        OracleInterface oracle = OracleInterface(oracleModule);
+
+        OtokenInterface otoken = OtokenInterface(_otoken);
+
+        address otokenUnderlyingAsset = otoken.underlyingAsset();
+        address otokenStrikeAsset = otoken.strikeAsset();
+        address otokenCollateralAsset = otoken.collateralAsset();
+        uint256 otokenExpiryTimestamp = otoken.expiryTimestamp();
+
+        bytes32 batch = keccak256(
+            abi.encode(otokenUnderlyingAsset, otokenStrikeAsset, otokenCollateralAsset, otokenExpiryTimestamp)
+        );
+
+        return oracle.isDisputePeriodOver(batch, otokenExpiryTimestamp);
+    }
 
     /**
-     * @notice Return a specific vault.
-     * @param _owner vault owner.
-     * @param _vaultId vault.
-     * @return Vault struct
+     * @notice get batch for a specific otoken address
+     * @dev batch is the hash of option underlying, strike, collateral assets and the expiry timestamp
+     * @param _otoken otoken address
+     * @return batch hash in bytes32
      */
-    // function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
-    // }
+    function getBatch(address _otoken) external view returns (bytes32) {
+        OtokenInterface otoken = OtokenInterface(_otoken);
+
+        return
+            keccak256(
+                abi.encode(
+                    otoken.underlyingAsset(),
+                    otoken.strikeAsset(),
+                    otoken.collateralAsset(),
+                    otoken.expiryTimestamp()
+                )
+            );
+    }
 
     /**
-     * @notice Checks if the sender is the operator of the owner’s account
-     * @param _owner The owner of the account
-     * @param _operator account operator
-     * @return true if it is an account operator, otherwise false
+     * @notice function to check if otoken is expired
+     * @param _otoken otoken address
+     * @return true if otoken is expired, else return false
      */
-    // function isOperator(address _owner, address _operator) public view returns (bool) {
-    // }
+    function isExpired(address _otoken) public view returns (bool) {
+        uint256 otokenExpiryTimestamp = OtokenInterface(_otoken).expiryTimestamp();
+
+        return now > otokenExpiryTimestamp;
+    }
+
+    /**
+     * @notice Return a specific vault.
+     * @param _owner account owner.
+     * @param _vaultId vault.
+     * @return Vault struct
+     */
+    function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
```

Comment:
Does this function need to be public or is internal enough ?

Reply:
Why internal ? can be public to and used for any case IMO

Reply:
is it used publicly? Just don't want to confuse users since this won't give the accurate collateral balances

Reply:
Yeah I think we need it public to show vault data right ?

Reply:
We need this. So users can easily use getVault and than getExcessMargin on their own.

Reply:
wait why do we need this, still don't understand

Reply:
when will users use the info about their vault? They really only care about how much actual collateral they have right, not colalteral that doesn't belong to them

Reply:
this will confuse developers more imo